test/Microsoft.AspNet.Mvc.IntegrationTests/BodyValidationIntegrationTests.cs

```diff
@@ -0,0 +1,263 @@
+                {
+                    BinderModelName = "CustomParameter",
+                    BindingSource = BindingSource.Body
+                },
+                ParameterType = typeof(Person)
+            };
+
+            var operationContext = ModelBindingTestHelper.GetOperationBindingContext();
+            var httpContext = operationContext.HttpContext;
+
+            ConfigureHttpRequest(httpContext.Request, "{ \"Id\":1234 }");
+            var modelState = new ModelStateDictionary();
+
+            // Act
+            var modelBindingResult = await argumentBinder.BindModelAsync(parameter, modelState, operationContext);
+
+            // Assert
+            Assert.NotNull(modelBindingResult);
+            Assert.True(modelBindingResult.IsModelSet);
+            var boundPerson = Assert.IsType<Person>(modelBindingResult.Model);
+            Assert.NotNull(boundPerson);
+            var key = Assert.Single(modelState.Keys);
+            Assert.Equal("Address", key);
+            Assert.False(modelState.IsValid);
+            Assert.Equal("The Address field is required.", modelState[key].Errors.Single().ErrorMessage);
+        }
+
+        private class Person4
+        {
+            [FromBody]
+            [Required]
+            public int Address { get; set; }
+        }
+
+        [Fact]
+        public async Task FromBodyAndRequiredOnValueTypeProperty_EmptyBody_AddsModelStateError()
+        {
+            // Arrange
+            var argumentBinder = ModelBindingTestHelper.GetArgumentBinder();
+            var parameter = new ParameterDescriptor()
+            {
+                Name = "Parameter1",
+                BindingInfo = new BindingInfo()
+                {
+                    BinderModelName = "CustomParameter",
+                },
+                ParameterType = typeof(Person4)
+            };
+
+            var operationContext = ModelBindingTestHelper.GetOperationBindingContext();
+            var httpContext = operationContext.HttpContext;
+            ConfigureHttpRequest(httpContext.Request, string.Empty);
+            var actionContext = httpContext.RequestServices.GetRequiredService<IScopedInstance<ActionContext>>().Value;
+            var modelState = actionContext.ModelState;
+
+            // Act
+            var modelBindingResult = await argumentBinder.BindModelAsync(parameter, modelState, operationContext);
+
+            // Assert
+            Assert.NotNull(modelBindingResult);
+            Assert.True(modelBindingResult.IsModelSet);
+            var boundPerson = Assert.IsType<Person4>(modelBindingResult.Model);
+            Assert.NotNull(boundPerson);
+            Assert.False(modelState.IsValid);
+
+            // The error with an empty key is a bug(#2416)  in our implementation which does not append the prefix and
+            // use that along with the path. The expected key here would be CustomParameter.Address.
+            var key = Assert.Single(modelState.Keys, k => k == "");
```

Comment:
inconsistency with previous tests makes this a bit harder to understand.  suggest instead the following where it's the second line which shows the bug.
``` c#
var key = Assert.Single(modelState.Keys);
Assert.Empty(key);
```